### PR TITLE
[docs/spec/auth] Clarify the Token auth workflow

### PR DESCRIPTION
**Pull request by jlhawn**
Clarify that the token server **should not** return authorization errors.

Clarify that the registry **should** return a 403 response status if the token is invalid.
(The registry does not currently meet this requirement - see https://github.com/docker/distribution/issues/967).

### CODE REVIEW

File: docs/spec/auth/token.md [hunk 1]
```diff
@@ -55,19 +55,20 @@ The new registry workflow is more like this:
    HTTP response with information on how to authenticate.
 3. The registry client makes a request to the authorization service for a
    signed JSON Web Token.
-4. The authorization service returns a token.
+4. The authorization service returns a token with a subset of the requested
```
Comment (stevvooe) [inline docs/spec/auth/token.md]: I don't think this should be restricted to a "subset". A superset of the permissions may be valid, as well. Perhaps, this should read:

    The authorization service returns an opaque token representing the available grants to the client.

File: docs/spec/auth/token.md [hunk 2]
```diff
@@ -132,45 +149,51 @@ the `WWW-Authenticate` header.
...
-the repository `samalba/my-app` hosted by the service `registry.docker.com`.
+The token server should first attempt to authenticate the client using any
+authentication credentials provided with the request. As of Docker 1.8, the
+registry client in the Docker Engine only supports Basic Authentication to
+these token servers. If an attempt to authenticate to the token sever fails,
+the token server should return a `401 Unauthorized` response indicating that
+the provided credentials are invalid.
+
+Whether the token server requires authentication is up to the policy of that
+access control provider. Some requested access may require authenticaiton to
```
Comment (stevvooe) [inline docs/spec/auth/token.md]: "authenticaiton"

**Comment by stevvooe** [comment]
There are a few nits that should be addressed before merging, but I'm providing a contingent LGTM.

File: docs/spec/auth/token.md (hunk 1 above)
Comment (jlhawn) [inline docs/spec/auth/token.md]: How about:

```The authorization service returns an opaque token representing the client's access to the requested resource.```

Comment (stevvooe) [inline docs/spec/auth/token.md]: :nose:

**Comment by stevvooe** [comment]
LGTM

File: docs/spec/auth/token.md [hunk 3]
```diff
@@ -55,19 +55,20 @@ The new registry workflow is more like this:
...
    signed JSON Web Token.
-4. The authorization service returns a token.
+4. The authorization service returns an opaque token representing the client's
+   authorized access.
 5. The client retries the original request with the token embedded in the
    request header.
-6. The Registry authorizes the client and begins the push/pull session as
-   usual. 
+6. The Registry authorizes the client by validating the token and the claim set
+   embeded within it and begins the push/pull session as usual.
```
Comment (RichardScothern) [inline docs/spec/auth/token.md]: s/embeded/embedded

File: docs/spec/auth/token.md [hunk 4]
```diff
@@ -132,45 +149,51 @@ the `WWW-Authenticate` header.
...
-For this example, the client makes an HTTP request to the following endpoint
-over TLS using a client certificate with the server being configured to allow a
-non-verified issuer during the handshake (i.e., a self-signed client cert is
-okay).
+For this example, the client makes an HTTP request to the following URL:
 
 ```
-GET /v2/token/?service=registry.docker.com&scope=repository:samalba/my-app:push&account=jlhawn HTTP/1.1
-Host: auth.docker.com
+https://auth.docker.io/token?service=registry.docker.io&scope=repository:samalba/my-app:pull,push HTTP/1.1
```
Comment (RichardScothern) [inline docs/spec/auth/token.md]: Remove the `HTTP/1.1` or add back the `GET` for consistency?

File: docs/spec/auth/token.md [hunk 5]
```diff
@@ -132,45 +149,51 @@ the `WWW-Authenticate` header.
...
-the repository `samalba/my-app` hosted by the service `registry.docker.com`.
+The token server should first attempt to authenticate the client using any
+authentication credentials provided with the request. As of Docker 1.8, the
+registry client in the Docker Engine only supports Basic Authentication to
+these token servers. If an attempt to authenticate to the token sever fails,
+the token server should return a `401 Unauthorized` response indicating that
+the provided credentials are invalid.
+
+Whether the token server requires authentication is up to the policy of that
+access control provider. Some requested access may require authentication to
```
Comment (RichardScothern) [inline docs/spec/auth/token.md]: 'requested access' -> 'requests'

File: docs/spec/auth/token.md [hunk 6]
```diff
@@ -132,45 +149,51 @@ the `WWW-Authenticate` header.
...
+access control list to determine whether the client has the requested scope. In
+this example request, if I have authenticated as user `jlhawn`, the token
+server will determine what access I have to the repository `samalba/my-app`
+hosted by the entity `registry.docker.io`.
+
+Once the token server has determined what access the client has to the
+resources requested in the `scope` parameter, it will take the intersection of
+the set of requested actions on each resource and the set of actions that the
+client has in fact been granted. If the client only has a subset of the
+requested access **it should not be considered an error** as it should not be
```
Comment (RichardScothern) [inline docs/spec/auth/token.md]: 'it should not be considered the responsibility' -> 'isn't considered the responsibility'

File: docs/spec/auth/token.md [hunk 7]
```diff
@@ -416,8 +442,16 @@ claim set within. The registry will:
...
-server.
+If any of these requirements are not met, the registry will return a
+`403 Forbidden` response to indicate that the token is invalid.
+
+> NOTE: it is only at this point in the workflow that an authorization error
+> may occur. The token server should *not* return errors when the user does
+> not have the requested authorization. Instead, the returned token should
+> indicate whatever of the requested scope the client does have (the
+> intersection of reuested and granted access). If the token does not supply
+> propper authorization then the registry will return the appropriate error.
```
Comment (RichardScothern) [inline docs/spec/auth/token.md]: propper -> proper

**Comment by RichardScothern** [comment]
A couple of minor points, but overall a good read.  LGTM

File: docs/spec/auth/token.md [hunk 8]
```diff
@@ -132,45 +149,51 @@ the `WWW-Authenticate` header.
...
+The token server should first attempt to authenticate the client using any
+authentication credentials provided with the request. As of Docker 1.8, the
+registry client in the Docker Engine only supports Basic Authentication to
+these token servers. If an attempt to authenticate to the token sever fails,
+the token server should return a `401 Unauthorized` response indicating that
+the provided credentials are invalid.
+
+Whether the token server requires authentication is up to the policy of that
+access control provider. Some requested access may require authentication to
+determine access (such us pushing or pulling a private repository) while some
```
Comment (thaJeztah) [inline docs/spec/auth/token.md]: s/such us/such as/

File: docs/spec/auth/token.md [hunk 9]
```diff
@@ -132,45 +149,51 @@ the `WWW-Authenticate` header.
...
+access may not (such as pulling from a public repository).
```
Comment (thaJeztah) [inline docs/spec/auth/token.md]: s/while some access may not/while others may not/

File: docs/spec/auth/token.md [hunk 10]
```diff
@@ -132,45 +149,51 @@ the `WWW-Authenticate` header.
...
+Once the token server has determined what access the client has to the
+resources requested in the `scope` parameter, it will take the intersection of
+the set of requested actions on each resource and the set of actions that the
+client has in fact been granted. If the client only has a subset of the
+requested access **it should not be considered an error** as it should not be
+considered the responsibility of the token server to indicate authorization
+errors as part of this workflow.
+
+Continuing with the example request, the token server will find that the
+client's set of granted access to the repository is `{pull, push}` which when
```
Comment (thaJeztah) [inline docs/spec/auth/token.md]: should `{pull, push}` be `["pull", "push"]` (here, and below?)

File: docs/spec/auth/token.md [hunk 11]
```diff
@@ -416,8 +442,16 @@ claim set within. The registry will:
...
 
-At no point in this process should the registry need to <em>call back</em> to
-the authorization server. If anything, it would only need to update a list of
-trusted public keys for verifying token signatures or use a separate API
-(still to be spec'd) to add/update resource records on the authorization
-server.
+If any of these requirements are not met, the registry will return a
+`403 Forbidden` response to indicate that the token is invalid.
+
+> NOTE: it is only at this point in the workflow that an authorization error
```
Comment (thaJeztah) [inline docs/spec/auth/token.md]: Nit; `NOTE:` -> `**Note**:`

File: docs/spec/auth/token.md [hunk 12]
```diff
@@ -416,8 +442,16 @@ claim set within. The registry will:
...
+> may occur. The token server should *not* return errors when the user does
+> not have the requested authorization. Instead, the returned token should
+> indicate whatever of the requested scope the client does have (the
+> intersection of reuested and granted access). If the token does not supply
```
Comment (thaJeztah) [inline docs/spec/auth/token.md]: s/reuested/requested/

File: docs/spec/auth/token.md (hunk 6 above)
Comment (thaJeztah) [inline docs/spec/auth/token.md]: Or even (more explicitly); "it is not the responsibility".

Also; "it should not be considered an error"; this give still room for interpretation; is it *allowed* to produce an error? If not, we should be explicit about that.

File: docs/spec/auth/token.md [hunk 13]
```diff
@@ -132,45 +149,51 @@ the `WWW-Authenticate` header.
...
-The server first inspects the client certificate to extract the subject key and
-lookup which account it is associated with. The client is now authenticated
-using that account.
-
-The server next searches its access control list for the account's access to
-the repository `samalba/my-app` hosted by the service `registry.docker.com`.
+The token server should first attempt to authenticate the client using any
+authentication credentials provided with the request. As of Docker 1.8, the
+registry client in the Docker Engine only supports Basic Authentication to
+these token servers. If an attempt to authenticate to the token sever fails,
```
Comment (thaJeztah) [inline docs/spec/auth/token.md]: s/the token sever/the token server/

**Comment by thaJeztah** [comment]
Thanks @jlhawn!

**Pull request merged**